ksqldb-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java

```diff
@@ -181,8 +190,10 @@ protected void uncaughtHandler(final Thread t, final Throwable e) {
           );
       queryStateListener.ifPresent(lis -> lis.onError(queryError));
       queryErrors.add(queryError);
-      LOG.error("Unhandled exception caught in streams thread {}. ({})", t.getName(), errorType, e);
+      LOG.error("Unhandled exception caught in streams thread {}. ({})", Thread.currentThread().getName(), errorType, e);
     }
+    retryEvent.backOff();
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
```

Comment:
for now we will always retry after a back off period

Reply:
Back off within the handler means we would tie-up that calling thread. Would this be a problem?

Reply:
No, because once the handler returns the thread will spawn a new one and then die. We can do this because the calling thread will be the one with the error

Reply:
Yes, I know this function is triggered as part of `StreamThread.runLoop` where we call `this.streamsUncaughtExceptionHandler.accept(e);`, and in which the caller thread is the dying stream thread. I'm just wondering if we should let the dying thread fail fast and use a scheduler with backoff to create a new thread instead of letting it to wait for a backoff time and then die --- of course, this is more complex than the current simple design.

If the motivation is to not triggering consecutive rebalances moving tasks around assuming the decision is to REPLACE_THREAD, I think that makes sense. But then it should not backoff for longer than poll interval since otherwise the thread would be kicked out of the group and rebalance triggered anyways.

Reply:
Ah. We can't let the thread die individually with the current streams functionality. I believe the max back off we set is 1 second by default so that shouldn't be a problem. Although it would bet best to have the max backoff time <  poll interval.

Reply:
Sounds good.

Reply:
The backoff starts at 15 seconds and goes up to 15 minutes. 

> and in which the caller thread is the dying stream thread. I'm just wondering if we should let the dying thread fail fast and use a scheduler with backoff to create a new thread instead of letting it to wait for a backoff time and then die --- of course, this is more complex than the current simple design.

I don't think we need another scheduler. Why not just have this stream thread clean itself up (close all clients), _then_ call the handler?

Reply:
If the decision was to shutdown the app, then sure we should just do that asap; but here the decision is to replace thread after backoff, then I think we do not need to set the max backoff to 15min since it is too long and would trigger a rebalance.

Reply:
The max back off would only happen after there has been several failures. in the first five minutes there will be four tires at 15s, 30s, 1m, 2m. If these all fail I think there should probably be a rebalance and while this thread gets up an runnning